src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7438,6 +7460,224 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
+            {
+                Debug.Assert(argIsRange);
+                // Look for Substring
+                var substring = (MethodSymbol)Compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Substring);
+                if (substring is { })
+                {
+                    patternIndexerAccess = new BoundIndexOrRangePatternIndexerAccess(
+                        syntax,
+                        receiverOpt,
+                        lengthOrCountProperty,
+                        substring,
+                        arguments[0],
+                        substring.ReturnType);
+                    cleanup();
+                    return true;
+                }
+            }
+            else
+            {
+                Debug.Assert(argIsRange);
+                // Look for `T Slice(int, int)` indexer
+
+                LookupMembersInType(
+                    lookupResult,
+                    receiverType,
+                    WellKnownMemberNames.SliceMethodName,
+                    arity: 0,
+                    basesBeingResolved: null,
+                    LookupOptions.Default,
+                    originalBinder: this,
+                    diagnose: false,
+                    ref useSiteDiagnostics);
+
+                if (lookupResult.IsMultiViable)
+                {
+                    foreach (var candidate in lookupResult.Symbols)
+                    {
+                        if (!candidate.IsStatic &&
+                            IsAccessible(candidate, ref useSiteDiagnostics) &&
+                            candidate is MethodSymbol method &&
+                            method.OriginalDefinition is var original &&
+                            original.ParameterCount == 2 &&
+                            isIntNotByRef(original.Parameters[0]) &&
+                            isIntNotByRef(original.Parameters[1]))
+                        {
+                            patternIndexerAccess = new BoundIndexOrRangePatternIndexerAccess(
+                                syntax,
+                                receiverOpt,
+                                lengthOrCountProperty,
+                                method,
+                                arguments[0],
+                                method.ReturnType);
+                            cleanup();
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            cleanup();
+            patternIndexerAccess = null;
+            return false;
+
+            bool isIntNotByRef(ParameterSymbol param)
+                => param.Type.SpecialType == SpecialType.System_Int32 &&
+                   param.RefKind == RefKind.None;
+
+            void cleanup()
+            {
+                lookupResult.Free();
+                useSiteDiagnostics = null;
+            }
+
+            bool lookupSatisfiesLengthPattern(out PropertySymbol valid)
```

Comment:
The containing method is pretty long here. Consider passing `lookupResult` and `useSiteDiagnostics` here to make the callsite implications a bit clearer.